File: scala/src/main/scala/banking/BankAccount.scala
package banking

final class BankAccount private (
    val openingBalance: Balance,
    val balance: Balance,
    val withdrawals: Vector[WithdrawalAmount]
):
  def withdraw(amount: WithdrawalAmount): Either[Withdraw.WithdrawalError, BankAccount] =
    Withdraw.withdraw(balance, amount).map { newBalance =>
      BankAccount.validated(openingBalance, newBalance, withdrawals :+ amount)
    }

object BankAccount:
  def open(openingBalance: Balance): BankAccount =
    validated(openingBalance, openingBalance, Vector.empty)

  private def validated(
      openingBalance: Balance,
      balance: Balance,
      withdrawals: Vector[WithdrawalAmount]
  ): BankAccount =
    val withdrawnTotal = withdrawals.foldLeft(0)(_ + _.value)
    assert(
      balance.value + withdrawnTotal == openingBalance.value,
      "account invariant failed: balance plus withdrawals must equal opening balance"
    )
    new BankAccount(openingBalance, balance, withdrawals)

File: kotlin/src/main/kotlin/banking/BankAccount.kt
package banking

class BankAccount(val openingBalance: Balance) {
    var balance: Balance = openingBalance
        private set

    private val mutableWithdrawals = mutableListOf<WithdrawalAmount>()
    val withdrawals: List<WithdrawalAmount>
        get() = mutableWithdrawals.toList()

    init {
        checkInvariant(balance, mutableWithdrawals)
    }

    fun withdraw(amount: WithdrawalAmount) {
        val candidateBalance = banking.withdraw(balance, amount)
        val candidateWithdrawals = mutableWithdrawals + amount

        // Validate the whole candidate state before changing this object.
        checkInvariant(candidateBalance, candidateWithdrawals)
        balance = candidateBalance
        mutableWithdrawals.add(amount)
    }

    private fun checkInvariant(
        candidateBalance: Balance,
        candidateWithdrawals: List<WithdrawalAmount>,
    ) {
        val withdrawnTotal = candidateWithdrawals.sumOf { it.value }
        check(candidateBalance.value + withdrawnTotal == openingBalance.value) {
            "account invariant failed: balance plus withdrawals must equal opening balance"
        }
    }
}
